packages/headless/src/api/search/date/date-format.ts

```diff
@@ -1,10 +1,14 @@
 import dayjs from 'dayjs';
 import customParseFormat from 'dayjs/plugin/customParseFormat';
+import minMax from 'dayjs/plugin/minMax';
+
 dayjs.extend(customParseFormat);
+dayjs.extend(minMax);
+
+const API_DATE_FORMAT = 'YYYY/MM/DD@HH:mm:ss';
 
 export function formatDateForSearchApi(date: dayjs.Dayjs) {
-  const DATE_FORMAT = 'YYYY/MM/DD@HH:mm:ss';
-  return date.format(DATE_FORMAT);
+  return dayjs.max([dayjs(new Date(0, 0, 0)), date]).format(API_DATE_FORMAT);
```

Comment:
Sure many customers won't have documents before `1899`, but I can see a law firm buying Coveo might want to index old cases?

It feels arbitrary to have a lower limit in headless, that does not exist in the search api or index.

Could you give us more context around the out of bounds issue? Perhaps there is an alternative approach we can use to solve the problem,

Reply:
The API (and index I guess) won't go lower than `1401/01/01@00:00:00`, if you send a lower value, that's what it will return.

Now if you go ridiculously low and send the API something like `-271769/02/06@17:44:36` you'll get a huge value back, higher than the end value -> `9999/01/01@00:59:59`

JS (if you play around with `new Date(...)`) will have trouble with values lower than 1899, and so does Day.js, which is why I picked that value, which is `new Date(0, 0, 0)`. I though, like you, that we propably won't get clients with dates older than this, but we could always workaround JS' limitation and go down to 1401, like the API/index

Reply:
> but we could always workaround JS' limitation and go down to 1401, like the API/index

So... sometimes, when a document is "bugged" in the index (ie: somehow, someone is able to push a document with no date in the index), the index will set a fallback value of `1401/01/01@00:00:00` (that's date `0` in C++ Boost library).

So, it's impossible to have a document with no date in the index, and it's impossible to have a document with a date lower than 1401/01/01@00:00:00.

Following that logic, we should be able to handle date as low as 1401/01/01@00:00:00 in Headless, because otherwise it means we would not be able to query for those "illogical date" document in the index. Which would be a problem.

So, it makes sense to add some boundaries for dates in general in Headless, if only to prevent some super weird side effect.

But it should follow the limits in the index IMO.

Reply:
Cool, thanks for the additional detail.

I think matching the api/index limit vs. headless restricting the possible range further client-side makes sense?

Also why not throw an error with a clear message, versus silently trying to recover? Especially for dates with a negative year, month, day, but even dates before 1401. It feels like that behaviour would be more consistent with other parts of the code-base using bueno?


Reporting a bug for the search api team would be helpful too, since protections we add in headless aren't available to people using Coveo's apis directly.